apiserver/common/networkingcommon/types.go

```diff
@@ -206,6 +206,26 @@ func networkDeviceToStateArgs(dev corenetwork.InterfaceInfo) state.LinkLayerDevi
 	}
 }
 
+// networkAddressStateArgsForHWAddr accommodates the fact that network
+// configuration is sometimes supplied with a duplicated device for each
+// address.
+// This is a normalisation that returns state args for all primary addresses
+// of interfaces with the input hardware address.
+func networkAddressStateArgsForHWAddr(devs corenetwork.InterfaceInfos, hwAddr string) []state.LinkLayerDeviceAddress {
+	var res []state.LinkLayerDeviceAddress
+
+	for _, dev := range devs.GetByHardwareAddress(hwAddr) {
```

Comment:
It seems that we resort to this grouping method quite a bit. Would it be worth investigating a potential change of InterfaceInfos into a type (e.g. map or struct with a slice and memoized map) that stores the interfaces in this form for easier access?

Reply:
I don't think we're ever dealing with enough of them at once to go to the trouble.